client/src/templates/Challenges/exam/show.tsx

```diff
@@ -362,28 +359,65 @@ class ShowExam extends Component<ShowExamProps, ShowExamState> {
   };
 
   runExam = () => {
-    // TODO: show loader
-    // TODO: fetch exam from server/database
-    const newExam = this.state.generatedExam.map(q => {
-      return { question: q.question, answer: null };
-    });
+    const {
+      data: {
+        challengeNode: {
+          challenge: { prerequisites }
+        }
+      },
+      completedChallenges,
+      createFlashMessage
+    } = this.props;
+
+    // check that they have all the prerequisites before they can start
+    let qualifiedForExam = true;
+    let incompleteChallenge = '';
+
+    for (let i = 0; i < prerequisites.length; i++) {
+      const prerequisiteChallenge = prerequisites[i];
+      const isPrerequisiteCompleted = completedChallenges.find(
+        (completedChallenge: CompletedChallenge) =>
+          prerequisiteChallenge.id === completedChallenge.id
+      );
+
+      if (!isPrerequisiteCompleted) {
+        qualifiedForExam = false;
+        incompleteChallenge = prerequisiteChallenge.title;
+        break;
+      }
+    }
 
-    this.timerInterval = setInterval(() => {
-      this.setState({
-        examTimeInSeconds: this.state.examTimeInSeconds + 1
+    if (!qualifiedForExam) {
+      createFlashMessage({
+        type: 'danger',
+        message: FlashMessages.MissingPrerequisite,
+        variables: { challengeTitle: incompleteChallenge }
+      });
+    } else {
+      // qualified to take the exam
+      // TODO: show loader
+      // TODO: fetch exam from server/database
+      const newExam = this.state.generatedExam.map(q => {
+        return { question: q.question, answer: null };
       });
-    }, 1000);
 
-    this.setState(
-      {
-        userExam: newExam
-      },
-      this.props.startExam
-    );
+      this.timerInterval = setInterval(() => {
+        this.setState({
+          examTimeInSeconds: this.state.examTimeInSeconds + 1
+        });
```

Comment:
```suggestion
        this.setState(state => ({
          examTimeInSeconds: state.examTimeInSeconds + 1
        }));
```
Since state updates are async, there's a risk that this.state is out of date by the time setState is called. The React docs talk about it [here](https://legacy.reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous)

Reply:
✅